dtm0/fop.c

```diff
@@ -307,55 +309,97 @@ M0_INTERNAL int m0_dtm0_logrec_update(struct m0_be_dtm0_log  *log,
 	return M0_RC(rc);
 }
 
-M0_INTERNAL int m0_dtm0_on_committed(struct m0_fom                *fom,
-				     const struct m0_dtm0_tx_desc *txd)
+M0_INTERNAL int m0_dtm0_on_committed(struct m0_fom            *fom,
+				     const struct m0_dtm0_tid *id)
 {
 	struct m0_dtm0_service *dtms = m0_dtm0_service_find(
-		fom->fo_service->rs_reqh);
-	struct m0_dtm0_tx_desc  msg = {};
+					fom->fo_service->rs_reqh);
+	struct m0_be_dtm0_log  *log = dtms->dos_log;
+	struct m0_dtm0_log_rec *rec;
+	struct m0_dtm0_tx_desc  txd;
 	int                     rc;
 	int                     i;
 
-	rc = m0_dtm0_tx_desc_copy(txd, &msg);
-	if (rc != 0)
-		return M0_ERR(rc);
 
+	M0_PRE(log != NULL);
 	/*
-	 * TODO: This change will be done by DTM0 log because
-	 * it should log the entry "pa == self" with PERSISTENT state
-	 * set.
+         * TODO: uncomment the following line after the connections code
+	 * is updated to detect and set volatile and persistent dtm0 service
+	 * instances. At present, the all2all test treats all connections as
+	 * volatile.
+	 * M0_PRE(log->dl_is_persistent);
 	 */
-	for (i = 0; i < msg.dtd_ps.dtp_nr; ++i) {
-		if (m0_fid_eq(&msg.dtd_ps.dtp_pa[i].p_fid,
-			      &dtms->dos_generic.rs_service_fid)) {
-			msg.dtd_ps.dtp_pa[i].p_state =
-				max_check(msg.dtd_ps.dtp_pa[i].p_state,
-					  (uint32_t) M0_DTPS_PERSISTENT);
-		}
-	}
+
+	M0_ENTRY();
+
+	m0_mutex_lock(&log->dl_lock);
+	/* Get the latest state of the log record. */
+	rec = m0_be_dtm0_log_find(log, id);
+	M0_ASSERT_INFO(rec != NULL, "Log record must be inserted into the log "
+		       "in cas_fom_tick().");
+	m0_dtm0_tx_desc_copy(&rec->dlr_txd, &txd);
+	m0_mutex_unlock(&log->dl_lock);
 
 	/* Notify the originator */
-	rc = m0_dtm0_send_msg(fom, DTM_PERSISTENT, &msg.dtd_id.dti_fid, &msg);
+	rc = m0_dtm0_send_msg(fom, DTM_PERSISTENT, &txd.dtd_id.dti_fid, &txd);
 
-	/* TODO: Send P msgs to the rest of the participants. */
-	m0_dtm0_tx_desc_fini(&msg);
+	if (rc != 0)
+		rc = M0_ERR_INFO(rc, "failed to send PERSISTENT msg to originator"
+			FID_F " -> " FID_F, FID_P(&dtms->dos_generic.  rs_service_fid),
+			FID_P(&txd.dtd_ps.dtp_pa->p_fid));
+	/* Send P msgs to the rest of the participants. */
+	for (i = 0; i < txd.dtd_ps.dtp_nr; ++i) {
+		if (m0_fid_eq(&txd.dtd_ps.dtp_pa[i].p_fid,
+			      &dtms->dos_generic.rs_service_fid))
+			continue;
+		rc = m0_dtm0_send_msg(fom, DTM_PERSISTENT,
+				       &txd.dtd_ps.dtp_pa[i].p_fid, &txd);
+		if (rc != 0)
+			rc = M0_ERR_INFO(rc, "failed to send PERSISTENT msg "
```

Comment:
We don't need a reassignment here, it is needed in situation like that

`		rc = M0_ERR_INFO(-ENOENT, "Failed to get pool version "FID_F,
				 FID_P(&cr->cr_pver));`

Reply:
Done. Fixed it.